### PR TITLE
FE/IN: Notifications

### PR DESCRIPTION
**Pull request by serafy0**
### What this PR does (required):
- added navbar
- added notifications component inside the navbar
- the notifications' menu will only show unread messages
- whenever a user opens the menu all items inside are marked as seen
- whenever a user creates a request it should also send a notification for the sitter. 

### Screenshots / Videos (required):
- Post at least one screenshot to show the feature in action
Screenshot 2021-09-08 at 18-25-58 Messenger

### Any information needed to test this feature (required):
- I've tried testing this by sending requests with postman then seeing if they will show up in the notification menu

### CODE REVIEW

File: client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  };
+  const [notifications, setNotifications] = useState<[Notification]>();
+  useEffect(() => {
+    async function fetchNotifications() {
+      if (loggedInUser) {
+        const fetchedNotifications = await getAllUnreadNotifications();
+        setNotifications(fetchedNotifications);
+      }
+    }
+    fetchNotifications();
```
Comment (ugur-akin) [inline client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx:45]: Wouldn't it be better if we could show old notifications to the user as well?

File: client/src/App.tsx [hunk 2]
```diff
@@ -20,6 +21,7 @@ function App(): JSX.Element {
         <SnackBarProvider>
           <AuthProvider>
             <SocketProvider>
+              <NavBar />
```
Comment (moffatethan) [inline client/src/App.tsx:24]: Good choice on the Navbar placement! 👍🏻

File: client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      <Button aria-controls="simple-menu" aria-haspopup="true" onClick={handleClick}>
+        <Badge variant="dot" color="primary" invisible={notifications && notifications.length < 1}>
+          <Typography variant="button">Notifications</Typography>
+        </Badge>
+      </Button>
+      {notifications && notifications.length >= 1 && (
+        <Menu
+          className={classes.menu}
+          elevation={0}
+          getContentAnchorEl={null}
```
Comment (ugur-akin) [inline client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx:59]: Is this prop necessary/does it contribute?

File: client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+          anchorOrigin={{
+            vertical: 'bottom',
+            horizontal: 'center',
+          }}
+          transformOrigin={{
+            vertical: 'top',
+            horizontal: 'center',
+          }}
+          id="simple-menu"
```
Comment (ugur-akin) [inline client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx:68]: Let's give this a more descriptive id?

File: client/src/components/NavBar/NavBar.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import AvatarDisplay from '../AvatarDisplay/AvatarDisplay';
+import { Link as routerLink } from 'react-router-dom';
+import NotificationsMenu from './NotificationsMenu/NotificationsMenu';
+export default function NavBar(): JSX.Element {
+  const classes = useStyles();
+  const { loggedInUser } = useAuth();
+
+  return (
+    <div className={classes.root}>
+      {loggedInUser && (
```
Comment (moffatethan) [inline client/src/components/NavBar/NavBar.tsx:13]: So, at some point, we will want to update this to conditionally show links based on account type. Sitters and users who needed their dog watched would need different contexts.

File: client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx (hunk 1 above)
Comment (moffatethan) [inline client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx:45]: Good work here!

File: client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+          open={Boolean(anchorEl)}
+          onClose={handleClose}
+        >
+          <Box display={{ xs: 'none', sm: 'block' }} className={classes.arrowUp} />
+          <Box borderTop={5}>
+            <Paper>
+              {notifications &&
+                notifications.map((notification, index) => (
+                  <MenuItem key={index} className={classes.notificationMenuItem} onClick={handleClose}>
+                    <ListItem>
```
Comment (ugur-akin) [inline client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx:80]: `MenuItem` wraps a `ListItem` component itself, we shouldn't need a second `ListItem` component here.

File: client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+export default function NotificationsMenu(): JSX.Element {
+  const classes = useStyles();
+  const { loggedInUser } = useAuth();
+
+  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
+
+  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
+    notifications?.map(async (notification) => {
+      await markAsRead(notification._id);
+    });
```
Comment (moffatethan) [inline client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx:29]: What I think would be even better, could we just send down a batch of notification ids to the controller and have the server perform batch processing on the notifications and update them through the database driver?

File: client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+              {notifications &&
+                notifications.map((notification, index) => (
+                  <MenuItem key={index} className={classes.notificationMenuItem} onClick={handleClose}>
+                    <ListItem>
+                      <ListItemAvatar>
+                        <Avatar
+                          src={
+                            notification.context
+                              ? notification.context.profilePhotoURL
+                              : 'https://res.cloudinary.com/dalisapxa/image/upload/v1630880229/DEV/logo_nywmrf.png'
```
Comment (ugur-akin) [inline client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx:86]: We should probably extract this URL to a variable. Or we can default to it somewhere appropriate too.

File: server/controllers/request.js [hunk 9]
```diff
@@ -49,8 +50,24 @@ exports.postRequest = asyncHandler(async (req, res, next) => {
...
   await Profile.updateOne({ _id: sitterId }, { $push: { requestsReceived: request._id } });
+  console.log(updatedUserProfile);
+  const currentUser = await User.findById(userId);
+
+  const newNotification = await Notification.create({
+    userId: sitterId,
+    notificationType: 'dog sitting',
+    title: `${
+      updatedUserProfile && updatedUserProfile.firstName ? updatedUserProfile.firstName : 'someone'
+    } has requested your service for ${parseInt(Math.abs(request.start - request.end) / 36e5)} hours`,
```
Comment (moffatethan) [inline server/controllers/request.js:66]: Anyway, this could be simplified, it's a bit hard to follow.

**Review by ugur-akin — COMMENTED** [review]
Good work! This looks good, I've left some minor comments to address.

File: server/models/Notification.js [hunk 10]
```diff
@@ -1,21 +1,24 @@
...
     title: { type: String, required: true },
     description: String,
     read: { type: Boolean, default: false },
     userId: {
       type: mongoose.Schema.Types.ObjectId,
-      ref: "user",
+      ref: 'user',
+    },
+    context: {
+      profilePhotoURL: String,
```
Comment (moffatethan) [inline server/models/Notification.js:18]: You are on the right track here with this, my idea for this was it's an object that can have anything in it. Mongoose has a Mixed data type, this aligns with what was initially envisioned here. Hope this helps

Comment (moffatethan) [reply server/models/Notification.js:18]: Essentially, we want this `context` property to be able to include anything which generalizes this for further support in notification types requiring various different contexts.

**Review by moffatethan — COMMENTED** [review]
Good work overall, along with Ugur's comments here are mine as well 👍🏻

File: client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx (hunk 3 above)
Comment (serafy0) [reply client/src/components/NavBar/NotificationsMenu/NotificationsMenu.tsx:57]: I kept them because whenever I tried to change or remove `getContentAnchorEl` or  `elevation`. the drop-down menu would cover up the navigation buttons

**Review by moffatethan — APPROVED** [review]
Good work!
